File: android/src/main/java/com/margelo/nitro/audioanvil/HybridAnvilFactory.kt
package com.margelo.nitro.audioanvil

import androidx.annotation.Keep
import com.facebook.proguard.annotations.DoNotStrip
import com.facebook.react.bridge.ReactApplicationContext
import com.margelo.nitro.NitroModules
import com.margelo.nitro.core.Promise

/** Autolinked root object. Creates recorders, answers permission questions, recovers orphaned sessions. */
@Keep
@DoNotStrip
class HybridAnvilFactory : HybridAnvilFactorySpec() {
  private val context: ReactApplicationContext
    get() = NitroModules.applicationContext ?: throw AnvilException(RecorderErrorCode.STATE, "No ApplicationContext set!")

  override fun createRecorder(config: RecorderConfig): Promise<HybridAnvilRecorderSpec> {
    return Promise.parallel<HybridAnvilRecorderSpec> {
      RecorderConfigValidator.validate(config)
      HybridAnvilRecorder(config)
    }
  }

  override fun getPermissionStatus(): PermissionStatus {
    return AnvilPermission.status(context, context.currentActivity)
  }

  override fun requestPermission(): Promise<PermissionStatus> {
    return AnvilPermission.request(context, context.currentActivity)
  }

  override fun discoverOrphanedRecordings(directory: String): Promise<Array<OrphanedRecording>> {
    return Promise.parallel {
      OrphanScanner.discover(AnvilPaths.directory(directory)).toTypedArray()
    }
  }

  override fun concatenate(segmentPaths: Array<String>, outputPath: String): Promise<RecordingSegment> {
    return Promise.parallel {
      WavConcatenator.concatenate(segmentPaths.map { AnvilPaths.directory(it) }, AnvilPaths.directory(outputPath))
    }
  }
}

File: ios/HybridAnvilFactory.swift
import AVFoundation
import Foundation
import NitroModules

/// Autolinked root object. Creates recorders, answers permission questions, recovers orphaned sessions.
final class HybridAnvilFactory: HybridAnvilFactorySpec {
  private static let queue = DispatchQueue(label: "ai.nitroaudio.anvil.factory")

  func createRecorder(config: RecorderConfig) throws -> Promise<any HybridAnvilRecorderSpec> {
    return Promise.parallel(Self.queue) { () throws -> any HybridAnvilRecorderSpec in
      try RecorderConfigValidator.validate(config)
      return HybridAnvilRecorder(config: config)
    }
  }

  func getPermissionStatus() throws -> AnvilPermissionStatus {
    return AVAudioSession.sharedInstance().anvilPermissionStatus
  }

  func requestPermission() throws -> Promise<AnvilPermissionStatus> {
    return Promise.async { await AVAudioSession.anvilRequestPermission() }
  }

  func discoverOrphanedRecordings(directory: String) throws -> Promise<[OrphanedRecording]> {
    return Promise.parallel(Self.queue) {
      try OrphanScanner.discover(directory: URL.anvilDirectory(directory))
    }
  }

  func concatenate(segmentPaths: [String], outputPath: String) throws -> Promise<RecordingSegment> {
    return Promise.parallel(Self.queue) {
      try WavConcatenator.concatenate(inputs: segmentPaths.map { URL.anvilFile($0) }, output: URL.anvilFile(outputPath))
    }
  }
}
